docker/docker_client.go

```diff
@@ -221,10 +221,87 @@ func CheckAuth(ctx context.Context, sCtx *types.SystemContext, username, passwor
 	}
 }
 
+// Data holds the output of both the v2 and v1 endpoints
+type Data struct {
+	// Repositories holds the results returned by the /v2/_catalog endpoint
+	Repositories []string `json:"repositories"`
+	// Results holds the results returned by the /v1/search endpoint
+	Results []Results `json:"results"`
+}
+
+// Results holds the information of each matching image returned by the v1 endpoint
+type Results struct {
+	Name        string `json:"name"`
+	Description string `json:"description"`
+	StarCount   int    `json:"star_count"`
+	IsTrusted   bool   `json:"is_trusted"`
+	IsAutomated bool   `json:"is_automated"`
+	IsOfficial  bool   `json:"is_official"`
+}
+
+// SearchRegistries queries a list of registries for a matching image
+func SearchRegistries(ctx context.Context, sCtx *types.SystemContext, registries []string, image, limit string) (map[string][]Results, error) {
+	searchResults := make(map[string][]Results)
+	for _, reg := range registries {
+		url := reg
+		if strings.Contains(reg, "docker") {
+			url = "index.docker.io"
+		}
+
+		logrus.Debugf("pinging v2 endpoint\n")
+		data, err := getRegistryData(ctx, sCtx, url, image, limit)
+		if err != nil {
+			fmt.Printf("couldn't search registry %q: %v", reg, err)
+			continue
+		}
+
+		if len(data.Results) > 0 {
+			searchResults[reg] = data.Results
+			continue
+		}
+
+		results := []Results{}
+		for _, repo := range data.Repositories {
+			if strings.Contains(repo, image) {
+				res := Results{
+					Name: repo,
+				}
+				results = append(results, res)
+			}
+		}
+		searchResults[reg] = results
+	}
+	return searchResults, nil
+}
+
+// getREgistryData talks to either the v2 or v1 endpoint to get the results of the search query
+func getRegistryData(ctx context.Context, sCtx *types.SystemContext, registry, image, limit string) (*Data, error) {
+	client, err := newDockerClientWithDetails(sCtx, registry, "", "", "", nil, "")
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating new docker client")
+	}
+	logrus.Debugf("pinging v2 endpoint\n")
+	resp, err := client.makeRequest(ctx, "GET", "/v2/_catalog", nil, nil, true)
+	if err != nil || resp.StatusCode != http.StatusOK {
+		logrus.Debugf("pinging v1 endpoint\n")
+		resp, err = client.makeRequest(ctx, "GET", "/v1/search?q="+image+"&n="+limit, nil, nil, true)
+		if err != nil || resp.StatusCode != http.StatusOK {
+			return nil, errors.Errorf("error getting response from endpoint")
```

Comment:
This should preserve the available data about the failure, i.e. either `err` or `res.StatusCode`.

Reply:
Fixed.

Reply:
AFAICT the error, and non-200 status, is still thrown away. Sure, the error message will have to be ugly if both v1 and v2 fail, but for debugging it would be nice to return something a bit useful.

Reply:
fixed. Logging the errors and response code now.

Reply:
Good idea, logging the two results is much clearer than trying to cram the data into an `error`.

Reply:
I changed it from logrus.Errorf to logrus.Debugf because the error was being logged always and I want it to only show if someone enables debugging. Hope that is acceptable.

Reply:
The principal case which I’m thinking about here is the user mistyping a hostname, or the network being down (guessing with absolutely no evidence that these are the most likely kinds of failures); in that case it would be nice to tell the user something useful about the failure by default.

Still, a `Debugf` is acceptable.

Reply:
Right, so using a Debugf to log any errors occurring, but if both the v1 and v2 endpoint fail an error is returned. The user can then enable debugging for further information.